### PR TITLE
Guard against nullptr dereference when checking concurrent solver status

### PR DESCRIPTION
**Pull request by vitor1001**
If dual simplex or barrier did not set their solution pointers (e.g. on early termination, error, or limit), dereferencing sol_dual_simplex_ptr or sol_barrier_ptr directly causes a null pointer dereference.

Full disclosure: done with the help of Gemini AI.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/cuopt/blob/HEAD/CONTRIBUTING.md).
- Testing
   - [ ] New or existing tests cover these changes
   - [ ] Added tests
   - [ ] Created an issue to follow-up
   - [ ] NA
- Documentation
   - [ ] The documentation is up to date with these changes
   - [ ] Added new documentation
   - [ ] NA

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
`sol_dual_simplex_ptr` is still unconditionally dereferenced in the later `PDLP ConcurrentLimit` branch. If dual simplex produced no result, the new status guard maps it to `CONCURRENT_LIMIT`, but this branch then does `*sol_dual_simplex_ptr` anyway and can still crash. Please guard/fallback there too and add a null-result `ConcurrentLimit` regression.

**Comment by vitor1001** [comment]
Great points, done!

**Comment by sylvesterkaczmarek** [comment]
Thanks, appreciate the quick turnaround. The null ConcurrentLimit dereference I flagged is fixed, and the new regression exercises the no-dual-simplex path. I see the separate RAII test-cleanup comment as well; that's independent of my concern.

File: cpp/tests/linear_programming/pdlp_test.cu [hunk 1]
```diff
@@ -56,6 +56,10 @@
 #include <utility>
 #include <vector>
 
+#ifdef _OPENMP
```
Comment (hlinsen) [inline cpp/tests/linear_programming/pdlp_test.cu]: OpenMP is a hard requirement you can skip the macro check

File: cpp/tests/linear_programming/pdlp_test.cu [hunk 2]
```diff
@@ -174,6 +178,59 @@ TEST(pdlp_class, concurrent_pdlp_exception_joins_worker_threads)
...
+
+  auto settings       = pdlp_solver_settings_t<int, double>{};
+  settings.method     = cuopt::mathematical_optimization::method_t::Concurrent;
+  settings.presolver  = cuopt::mathematical_optimization::presolver_t::None;
+  settings.inside_mip = true;
+
+  // inside_mip skips dual simplex. Setting threads to 1 ensures barrier is also disabled
+  // (< CUOPT_CONCURRENT_LP_BARRIER_REQUIRED_THREAD_COUNT), leaving both sol_dual_simplex_ptr
+  // and sol_barrier_ptr null.
+#ifdef _OPENMP
```
Comment (hlinsen) [inline cpp/tests/linear_programming/pdlp_test.cu]: same here

File: cpp/tests/linear_programming/pdlp_test.cu [hunk 3]
```diff
@@ -174,6 +178,59 @@ TEST(pdlp_class, concurrent_pdlp_exception_joins_worker_threads)
...
+  const int prev_threads = omp_get_max_threads();
+  omp_set_num_threads(1);
+#endif
```
[the remark this answers is not in the document]
Comment (hlinsen) [reply cpp/tests/linear_programming/pdlp_test.cu]: Other exceptions will cause the test to fail and CI to fail.

[the remark this answers is not in the document]
Comment (hlinsen) [reply cpp/tests/linear_programming/pdlp_test.cu]: @vitor1001 This is only an edge case, but we already have `cuopt::scope_guard` to make the restoration exception safe. Could you use it here to restore the OpenMP setting? Thanks for the PR!
```
#include <utilities/scope_guard.hpp>

const int prev_threads = omp_get_max_threads();
omp_set_num_threads(1);
const cuopt::scope_guard restore_threads{
  [prev_threads] { omp_set_num_threads(prev_threads); }};
```

File: cpp/tests/linear_programming/pdlp_test.cu (hunk 2 above)
Comment (vitor1001) [reply cpp/tests/linear_programming/pdlp_test.cu]: Agreed, done!

File: cpp/tests/linear_programming/pdlp_test.cu (hunk 1 above)
Comment (vitor1001) [reply cpp/tests/linear_programming/pdlp_test.cu]: Agreed, done!

File: cpp/tests/linear_programming/pdlp_test.cu (hunk 3 above)
[the remark this answers is not in the document]
Comment (vitor1001) [reply cpp/tests/linear_programming/pdlp_test.cu]: Good catch, done as suggested.

**Review by hlinsen — APPROVED** [review]
Thanks @vitor1001!
